R/cnd.R

```diff
@@ -279,65 +279,135 @@ muffle <- function(...) NULL
+#' [base::warning()] and [base::message()], but make it easy to supply
+#' condition metadata:
+#'
+#' * Supply `type` to create a classed condition. Typed conditions can
+#'   be captured or handled selectively, allowing for finer-grained
+#'   error handling.
+#'
+#' * Supply metadata with named `...` arguments. This data will be
+#'   stored in the condition object and can be examined by handlers,
+#'   print methods, etc.
+#'
+#'
+#' @section Call trace:
+#'
+#' Unlike `stop()` and `warning()`, these functions don't include call
+#' information by default. This saves you from typing `call. = FALSE`
+#' and produces cleaner error messages.
+#'
+#'
+#' @section Mufflable conditions:
+#'
+#' Signalling a condition with `inform()` or `warn()` causes a message
+#' to be displayed in the console. These messages can be muffled with
+#' [base::suppressMessages()] or [base::suppressWarnings()].
+#'
+#'
+#' @section Lifecycle:
+#'
+#' These functions were changed in rlang 0.3.0 to take condition
+#' metadata with `...`. Consequently:
+#'
+#' * All arguments were renamed to be prefixed with a dot.
+#' * `.call` (previously `call`) can no longer be passed positionally.
+#'
+#' @param .msg The message to display.
+#' @param .type Subclass of the condition.
+#' @param ... Additional data to be stored in the condition object.
+#' @param .call Whether to display the call. If a number `n`, the call
 #'   is taken from the nth frame on the [call stack][call_stack].
+#' @param msg,type,call These arguments were renamed to be prefixed
+#'   with a dot and are deprecated as of rlang 0.3.0.
+#'
 #' @export
-abort <- function(msg, type = NULL, call = FALSE) {
-  cnd <- error_cnd(type, .msg = msg, .call = cnd_call(call))
-  if (!is_false(call)) {
+#' @examples
+#' # These examples are guarded to avoid throwing errors
+#' if (FALSE) {
+#'
+#' # Signal an error with a message just like stop():
+#' abort("Something bad happened")
+#'
+#' # Give a class to the error:
+#' abort("Something bad happened", "somepkg_bad_error")
+#'
+#' # This will allow your users to handle the error selectively
+#' tryCatch(
+#'   somepkg_function(),
+#'   somepkg_bad_error = function(err) {
+#'     cnd_warn(err) # Demote the error to a warning
+#'     NA            # Return an alternative value
+#'   }
+#' )
+#'
+#' # You can also specify metadata that will be stored in the condition:
+#' abort("Something bad happened", "somepkg_bad_error", data = 1:10)
+#'
+#' # This data can then be consulted by user handlers:
+#' tryCatch(
+#'   somepkg_function(),
+#'   somepkg_bad_error = function(err) {
+#'     # Compute an alternative return value with the data:
+#'     recover_error(err$data)
+#'   }
+#' )
+#'
+#' }
+abort <- function(.msg, .type = NULL, ..., .call = FALSE, msg, type, call) {
+  validate_signal_args(msg, type, call)
+
+  cnd <- error_cnd(.type, ..., .msg = .msg, .call = cnd_call(.call))
+  if (!is_false(.call)) {
```

Comment:
We really should always be assigning something to the `call` component to respect the fields of the parent class.

Reply:
This would cause the call to be printed on error. Am I missing something?

Reply:
We'd assign `NULL` when `.call` is `FALSE`

Reply:
Good point, I'll change this separately the ctors need a little work.